components/budget_increment_analysis.py

```diff
@@ -0,0 +1,249 @@
+import numpy as np
+import pandas as pd
+import plotly.graph_objects as go
+from constants import FUNC_COLORS
+from utils import (
+    filter_country_sort_year,
+    empty_plot,
+    generate_error_prompt,
+    calculate_cagr,
+)
+
+DEFAULT_VISIBLE_CATEGORIES = [
+    "Health",
+    "Education",
+    "General public services",
+    "Overall budget",
+]
+
+NUM_YEARS = 5
+
+def render_fig_and_narrative(data, country, exp_type):
+    country_budget_changes_df = pd.DataFrame(data["expenditure_by_country_func_year"])
+    country_budget_changes_df = filter_country_sort_year(
+        country_budget_changes_df, country
+    )
+    country_budget_changes_df = country_budget_changes_df[
+        country_budget_changes_df["expenditure"].notna()
+        & (round(country_budget_changes_df["expenditure"]) != 0)
+    ]
+
+    overall_budget_df = country_budget_changes_df.groupby(
+        ["country_name", "year"], as_index=False
+    ).agg(
+        {
+            "expenditure": "sum",
+            "domestic_funded_budget": "sum",
+            "real_domestic_funded_budget": "sum",
+        }
+    )
+    overall_budget_df["func"] = "Overall budget"
+    country_budget_changes_df = pd.concat(
+        [country_budget_changes_df, overall_budget_df], ignore_index=True
+    )
+
+    country_budget_changes_df = country_budget_changes_df.sort_values(
+        ["country_name", "func", "year"]
+    )
+
+    for col in ["domestic_funded_budget", "real_domestic_funded_budget"]:
+        prev_col = f"prev_{col}"
+        yoy_col = f"yoy_{col}"
+        country_budget_changes_df[prev_col] = country_budget_changes_df.groupby(
+            ["country_name", "func"]
+        )[col].shift(1)
+        country_budget_changes_df[yoy_col] = (
+            (country_budget_changes_df[col] - country_budget_changes_df[prev_col])
+            / country_budget_changes_df[prev_col]
+        ) * 100
+
+    country_budget_changes_df.drop(
+        columns=[
+            "prev_domestic_funded_budget",
+            "prev_real_domestic_funded_budget",
+        ],
+        inplace=True,
+    )
+
+    available_years = sorted(country_budget_changes_df["year"].unique())
+    end_year = available_years[-1]
+    if len(available_years) > NUM_YEARS:
+        selected_years = available_years[-NUM_YEARS:]
+    else:
+        selected_years = available_years
+    start_year = selected_years[0]
+
+    country_budget_changes_df = country_budget_changes_df[
+        country_budget_changes_df["year"].isin(selected_years)
+    ]
+
+    num_years = len(selected_years)
+
+    # TODO Sandeep: remove after gap years are handled
```

Comment:
@bhupatiraju please handle gaps years in yoy calculation, corresponding plotting, and cagr calculation.